// @algorandfoundation/algokit-utils/types/algorand-client.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var types_accountManager = require('./account-manager.js');
var types_algorandClientTransactionCreator = require('./algorand-client-transaction-creator.js');
var types_algorandClientTransactionSender = require('./algorand-client-transaction-sender.js');
var types_appDeployer = require('./app-deployer.js');
var types_appManager = require('./app-manager.js');
var types_assetManager = require('./asset-manager.js');
var types_clientManager = require('./client-manager.js');
var types_composer = require('./composer.js');

/**
 * A client that brokers easy access to Algorand functionality.
 */
class AlgorandClient {
    constructor(config) {
        this._cachedSuggestedParamsTimeout = 3000; // three seconds
        this._defaultValidityWindow = undefined;
        this._clientManager = new types_clientManager.ClientManager(config, this);
        this._accountManager = new types_accountManager.AccountManager(this._clientManager);
        this._appManager = new types_appManager.AppManager(this._clientManager.algod);
        this._assetManager = new types_assetManager.AssetManager(this._clientManager.algod, () => this.newGroup());
        this._transactionSender = new types_algorandClientTransactionSender.AlgorandClientTransactionSender(() => this.newGroup(), this._assetManager, this._appManager);
        this._transactionCreator = new types_algorandClientTransactionCreator.AlgorandClientTransactionCreator(() => this.newGroup());
        this._appDeployer = new types_appDeployer.AppDeployer(this._appManager, this._transactionSender, this._clientManager.indexerIfPresent);
    }
    /**
     * Sets the default validity window for transactions.
     * @param validityWindow The number of rounds between the first and last valid rounds
     * @returns The `AlgorandClient` so method calls can be chained
     */
    setDefaultValidityWindow(validityWindow) {
        this._defaultValidityWindow = validityWindow;
        return this;
    }
    /**
     * Sets the default signer to use if no other signer is specified.
     * @param signer The signer to use, either a `TransactionSigner` or a `TransactionSignerAccount`
     * @returns The `AlgorandClient` so method calls can be chained
     */
    setDefaultSigner(signer) {
        this._accountManager.setDefaultSigner(signer);
        return this;
    }
    /**
     * Tracks the given account (object that encapsulates an address and a signer) for later signing.
     * @param account The account to register, which can be a `TransactionSignerAccount` or
     *  a `algosdk.Account`, `algosdk.LogicSigAccount`, `SigningAccount` or `MultisigAccount`
     * @example
     * ```typescript
     * const accountManager = AlgorandClient.mainnet()
     *  .setSignerFromAccount(algosdk.generateAccount())
     *  .setSignerFromAccount(new algosdk.LogicSigAccount(program, args))
     *  .setSignerFromAccount(new SigningAccount(mnemonic, sender))
     *  .setSignerFromAccount(new MultisigAccount({version: 1, threshold: 1, addrs: ["ADDRESS1...", "ADDRESS2..."]}, [account1, account2]))
     *  .setSignerFromAccount({addr: "SENDERADDRESS", signer: transactionSigner})
     * ```
     * @returns The `AlgorandClient` so method calls can be chained
     */
    setSignerFromAccount(account) {
        this._accountManager.setSignerFromAccount(account);
        return this;
    }
    /**
     * Tracks the given signer against the given sender for later signing.
     * @param sender The sender address to use this signer for
     * @param signer The signer to sign transactions with for the given sender
     * @returns The `AlgorandClient` so method calls can be chained
     */
    setSigner(sender, signer) {
        this._accountManager.setSigner(sender, signer);
        return this;
    }
    /**
     * Sets a cache value to use for suggested transaction params.
     * @param suggestedParams The suggested params to use
     * @param until A date until which to cache, or if not specified then the timeout is used
     * @returns The `AlgorandClient` so method calls can be chained
     */
    setSuggestedParamsCache(suggestedParams, until) {
        this._cachedSuggestedParams = suggestedParams;
        this._cachedSuggestedParamsExpiry = until ?? new Date(+new Date() + this._cachedSuggestedParamsTimeout);
        return this;
    }
    /**
     * Sets the timeout for caching suggested params.
     * @param timeout The timeout in milliseconds
     * @returns The `AlgorandClient` so method calls can be chained
     */
    setSuggestedParamsCacheTimeout(timeout) {
        this._cachedSuggestedParamsTimeout = timeout;
        return this;
    }
    /** Get suggested params for a transaction (either cached or from algod if the cache is stale or empty) */
    async getSuggestedParams() {
        if (this._cachedSuggestedParams && (!this._cachedSuggestedParamsExpiry || this._cachedSuggestedParamsExpiry > new Date())) {
            return {
                ...this._cachedSuggestedParams,
            };
        }
        this._cachedSuggestedParams = await this._clientManager.algod.getTransactionParams().do();
        this._cachedSuggestedParamsExpiry = new Date(new Date().getTime() + this._cachedSuggestedParamsTimeout);
        return {
            ...this._cachedSuggestedParams,
        };
    }
    /** Get clients, including algosdk clients and app clients. */
    get client() {
        return this._clientManager;
    }
    /** Get or create accounts that can sign transactions. */
    get account() {
        return this._accountManager;
    }
    /** Methods for interacting with assets. */
    get asset() {
        return this._assetManager;
    }
    /** Methods for interacting with apps. */
    get app() {
        return this._appManager;
    }
    /** Methods for deploying apps and managing app deployment metadata. */
    get appDeployer() {
        return this._appDeployer;
    }
    /** Start a new `AlgoKitComposer` transaction group */
    newGroup() {
        return new types_composer.default({
            algod: this.client.algod,
            getSigner: (addr) => this.account.getSigner(addr),
            getSuggestedParams: () => this.getSuggestedParams(),
            defaultValidityWindow: this._defaultValidityWindow,
            appManager: this._appManager,
        });
    }
    /**
     * Methods for sending a transaction.
     */
    get send() {
        return this._transactionSender;
    }
    /**
     * Methods for creating a transaction.
     */
    get createTransaction() {
        return this._transactionCreator;
    }
    // Static methods to create an `AlgorandClient`
    /**
     * Returns an `AlgorandClient` pointing at default LocalNet ports and API token.
     * @returns The `AlgorandClient`
     */
    static defaultLocalNet() {
        return new AlgorandClient({
            algodConfig: types_clientManager.ClientManager.getDefaultLocalNetConfig('algod'),
            indexerConfig: types_clientManager.ClientManager.getDefaultLocalNetConfig('indexer'),
            kmdConfig: types_clientManager.ClientManager.getDefaultLocalNetConfig('kmd'),
        });
    }
    /**
     * Returns an `AlgorandClient` pointing at TestNet using AlgoNode.
     * @returns The `AlgorandClient`
     */
    static testNet() {
        return new AlgorandClient({
            algodConfig: types_clientManager.ClientManager.getAlgoNodeConfig('testnet', 'algod'),
            indexerConfig: types_clientManager.ClientManager.getAlgoNodeConfig('testnet', 'indexer'),
            kmdConfig: undefined,
        });
    }
    /**
     * Returns an `AlgorandClient` pointing at MainNet using AlgoNode.
     * @returns The `AlgorandClient`
     */
    static mainNet() {
        return new AlgorandClient({
            algodConfig: types_clientManager.ClientManager.getAlgoNodeConfig('mainnet', 'algod'),
            indexerConfig: types_clientManager.ClientManager.getAlgoNodeConfig('mainnet', 'indexer'),
            kmdConfig: undefined,
        });
    }
    /**
     * Returns an `AlgorandClient` pointing to the given client(s).
     * @param clients The clients to use
     * @returns The `AlgorandClient`
     */
    static fromClients(clients) {
        return new AlgorandClient(clients);
    }
    /**
     * Returns an `AlgorandClient` loading the configuration from environment variables.
     *
     * Retrieve configurations from environment variables when defined or get default LocalNet configuration if they aren't defined.
     *
     * Expects to be called from a Node.js environment.
     *
     * If `process.env.ALGOD_SERVER` is defined it will use that along with optional `process.env.ALGOD_PORT` and `process.env.ALGOD_TOKEN`.
     *
     * If `process.env.INDEXER_SERVER` is defined it will use that along with optional `process.env.INDEXER_PORT` and `process.env.INDEXER_TOKEN`.
     *
     * If either aren't defined it will use the default LocalNet config.
     *
     * It will return a KMD configuration that uses `process.env.KMD_PORT` (or port 4002) if `process.env.ALGOD_SERVER` is defined,
     * otherwise it will use the default LocalNet config unless it detects testnet or mainnet.
     * @returns The `AlgorandClient`
     */
    static fromEnvironment() {
        return new AlgorandClient(types_clientManager.ClientManager.getConfigFromEnvironmentOrLocalNet());
    }
    /**
     * Returns an `AlgorandClient` from the given config.
     * @param config The config to use
     * @returns The `AlgorandClient`
     */
    static fromConfig(config) {
        return new AlgorandClient(config);
    }
}

exports.AlgorandClient = AlgorandClient;
exports.default = AlgorandClient;
